devito/passes/iet/misc.py

```diff
@@ -104,9 +113,88 @@ def relax_incr_dimensions(iet, **kwargs):
 
+            # Step 1: Interval offset care
+            # Here we create two dictionaries to keep track of offsets incurring in
+            # 'i' and 'i.root' iteration
+            min_map = {}  # For the Iteration minimum
+            root_min = roots_min[i.dim.root] + i.symbolic_min - i.dim.symbolic_min
+            min_map[i.dim.root.symbolic_min] = root_min
+            min_map[i.dim.symbolic_min] = i.symbolic_min
+
+            max_map = {}  # For the Iteration maximum
             root_max = roots_max[i.dim.root] + i.symbolic_max - i.dim.symbolic_max
-            iter_max = evalrel(min, [i.symbolic_max, root_max])
-            mapper[i] = i._rebuild(limits=(i.symbolic_min, iter_max, i.step))
+            max_map[i.dim.root.symbolic_max] = root_max
+            max_map[i.dim.symbolic_max] = i.symbolic_max
+
+            # Step 2: Generate assumptions deriving from hierarchical blocking. Assuming
+            # lower blocking levels perfectly fit within outer levels, we use depth of
+            # IncrDimensions as a rule to generate assumptions such as:
+            # e.g. for 2-level blocking [x (inner Incr), x0_blk1, x0_blk0, x (outer root)]
+            # [x0_blk1 > x0_blk0, x0_blk1_size > x0_blk0_size]
+            # `defines` keeps all the ancestors and descendants of `i.dim`.
+            defines = sorted(i.dim._defines, key=lambda x: (not x.is_Incr or -x._depth))
+
+            # Collect assumptions based on hierarchy
+            acs = [j for j in defines if j is not i.dim and j.is_Incr]
+            assumptions = []
+            assumptions.extend(chain.from_iterable((Gt(j.step, k.step), Gt(j, k)) for j, k
+                               in product(acs, acs) if j._depth > k._depth))
+
+            # Step 3: Reduce defines min/max candidates using assumptions on `defines`
+            defmin = reduce_relation(max, defines, assumptions)
+            defmax = reduce_relation(min, defines, assumptions)
+
+            # Collect symbolic_min, symbolic_max from defmin/defmax lists
+            defmin = [j.symbolic_min for j in defmin]
+            defmax = [j.symbolic_max for j in defmax]
+
+            # Drop duplicates from defmin/defmax and keep them Ordered
+            defmin = list(OrderedDict.fromkeys(defmin))
+            defmax = list(OrderedDict.fromkeys(defmax))
+
+            # Step 4: Drop a candidate if it is the symbolic_min/max of symbol candidates
+            for k in defmin.copy():  # TOFIX: Should not use a copy I guess?
+                try:
+                    if k.symbolic_min in defmin:
+                        defmin.remove(k.symbolic_min)
+                except:
+                    pass
+
+            for k in defmax.copy():
+                try:
+                    if k.symbolic_max in defmax:
+                        defmax.remove(k.symbolic_max)
+                except:
+                    pass
+
+            # At this point the candidates for the start and end of the `Iteration`
```

Comment:
Writing comments is an art (and I'm bad at it)... but isn't this way too long?

Reply:
sometimes *tiny* examples are enough

Reply:
ping

Reply:
Improved